Introduce MallocSpan, which is similar to MallocPtr but also keeps track of the size

cdumez opened the pull request:
#### 96ca30d7d7d9fe84af045d94ce4b1a4482a36c69
<pre>
Introduce MallocSpan, which is similar to MallocPtr but also keeps track of the size
<a href="https://bugs.webkit.org/show_bug.cgi?id=282341">https://bugs.webkit.org/show_bug.cgi?id=282341</a>

Reviewed by Geoffrey Garen, Michael Catanzaro, and Adrian Perez de Castro.

Introduce MallocSpan, which is similar to MallocPtr but also keeps track of the size,
for extra safety.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/MallocSpan.h: Added.
(WTF::MallocSpan::MallocSpan):
(WTF::MallocSpan::~MallocSpan):
(WTF::MallocSpan::operator=):
(WTF::MallocSpan::swap):
(WTF::MallocSpan::span const):
(WTF::MallocSpan::mutableSpan):
(WTF::MallocSpan::leakSpan):
(WTF::MallocSpan::malloc):
(WTF::MallocSpan::zeroedMalloc):
(WTF::MallocSpan::tryMalloc):
(WTF::MallocSpan::tryZeroedMalloc):
(WTF::MallocSpan::realloc):
(WTF::adoptMallocSpan):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanConstCast):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::HistoryEntryDataEncoder):
(WebKit::HistoryEntryDataEncoder::finishEncoding):
(WebKit::HistoryEntryDataEncoder::growCapacity):
(WebKit::HistoryEntryDataEncoder::capacity const):
(WebKit::HistoryEntryDataEncoder::mutableBuffer):
(WebKit::HistoryEntryDataEncoder::buffer const):
(WebKit::encodeSessionHistoryEntryData):

Canonical link: <a href="https://commits.webkit.org/285930@main">https://commits.webkit.org/285930@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21faaeedcc4540f6d654e092174aaeddefce3f84

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78590 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25454 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62778 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1430 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58339 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16683 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77283 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/48503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38748 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45436 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21339 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23787 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67353 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66898 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21685 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80109 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73474 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1533 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/867 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66642 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1678 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63860 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65916 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16352 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9856 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8011 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95255 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1497 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20917 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1526 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1514 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1533 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -0,0 +1,118 @@
+
+    std::span<const T> span() const { return spanConstCast<const T>(m_span); }
+    std::span<T> mutableSpan() { return m_span; }
+    std::span<T> leakSpan() { return std::exchange(m_span, std::span<T>()); }
```
Hm, I considered suggesting that this should be called `leakData` because the std::span itself is notably not leaked. But then the name does not parallel the other two functions `span` and `mutableSpan`, and surely those should not be renamed. So I suppose it's probably best as-is.

aperezdc replied:
How about `releaseSpan()`? As it is “released” from being managed by the `MallocSpan` that contains it. There are already other uses of the “release” verb so I suppose it might work.

cdumez replied:
I think "leak" is the right term here as it is leaking memory and the caller will have to free it manually. We don't usually release for this kind of thing.
There is a lot of precedent for "leak": `MallocPtr::leakPtr()`, `RefPtr::leakRef()`, `Ref::leakRef()`, `ConnectionHandler::leakSendRight()`

We have the concept of `release` in WebKit but it usually doesn't leak memory, e.g. `RefPtr::releaseNonNull()`, which returns a Ref.

geoffreygaren replied:
Can we add WARN_UNUSED_RETURN like we do for other leak functions?

geoffreygaren reviewed (approved):
r=me

geoffreygaren commented on Source/WTF/wtf/MallocSpan.h:
```diff
@@ -0,0 +1,118 @@
+
+    ~MallocSpan()
+    {
+        Malloc::free(m_span.data());
```
Hmmm... will default-constructed or moved-from MallocSpan end up freeing nullptr? I believe free of nullptr is OK in FastMalloc but not other mallocs.

mcatanzaro replied:
You've been working with C++ for so long that you've forgotten that free() is guaranteed to do nothing when you pass NULL. It's fine, at least according to the various standards bodies.

cdumez replied:
MallocPtr was already relying on this behavior BTW: 
```
    ~MallocPtr()
    {
        Malloc::free(m_ptr);
    }
```
and `m_ptr` could be null. So this is not a new requirement on the Allocator class template parameter.

cdumez committed:
Introduce MallocSpan, which is similar to MallocPtr but also keeps track of the size
https://bugs.webkit.org/show_bug.cgi?id=282341

Reviewed by Geoffrey Garen, Michael Catanzaro, and Adrian Perez de Castro.

Introduce MallocSpan, which is similar to MallocPtr but also keeps track of the size,
for extra safety.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/MallocSpan.h: Added.
(WTF::MallocSpan::MallocSpan):
(WTF::MallocSpan::~MallocSpan):
(WTF::MallocSpan::operator=):
(WTF::MallocSpan::swap):
(WTF::MallocSpan::span const):
(WTF::MallocSpan::mutableSpan):
(WTF::MallocSpan::leakSpan):
(WTF::MallocSpan::malloc):
(WTF::MallocSpan::zeroedMalloc):
(WTF::MallocSpan::tryMalloc):
(WTF::MallocSpan::tryZeroedMalloc):
(WTF::MallocSpan::realloc):
(WTF::adoptMallocSpan):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanConstCast):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::HistoryEntryDataEncoder):
(WebKit::HistoryEntryDataEncoder::finishEncoding):
(WebKit::HistoryEntryDataEncoder::growCapacity):
(WebKit::HistoryEntryDataEncoder::capacity const):
(WebKit::HistoryEntryDataEncoder::mutableBuffer):
(WebKit::HistoryEntryDataEncoder::buffer const):
(WebKit::encodeSessionHistoryEntryData):

Canonical link: https://commits.webkit.org/285930@main
